### PR TITLE
Make `HistCutMatrix::Init' be aware of groups.

### PR DESCRIPTION
**Pull request by trivialfis**
* Add checks for group size.

Address segfault in  #4098,  error in quantile is not resolved.

### CODE REVIEW

**Comment by hcho3** [comment]
@trivialfis Yes, to clarify, weights in ranking objects will have to be per-group basis. Does this PR make this assumption? For the quantile sketch, we can replicate the group weight to cover all data points in each group.

**Comment by trivialfis** [comment]
@hcho3 This PR accepts both **per-group** and **per-instance** until the last commit `see failed java tests`.  I don't know how long does it take to get Travis running.

There was a test in jvm-package scala that inputs different number of weights and groups:

https://travis-ci.org/dmlc/xgboost/jobs/490537869#L1636

Please ignore the incorrect check, it's fixed in later commits.  But the showed error message seems to support **per-instance** weighting.

**Comment by trivialfis** [comment]
@hcho3 Ha, there it's:
https://travis-ci.org/dmlc/xgboost/jobs/490661939#L1705

So it's a buggy test and I should remove/fix it?

**Comment by hcho3** [comment]
@trivialfis Yes, and we'll need to be clear about the use of per-group weights. It really doesn't make sense to give different weights to data rows in the same query session, since only the relative ordering of the data rows are important (hence learning-to-rank).

**Comment by trivialfis** [comment]
@hcho3 Thanks for the explanation.  That goes in my reading list.

**Comment by hcho3** [comment]
@trivialfis A good place to start is to learn what the ranking metric (e.g. NDCG) does. This reading helped me: http://dalelane.co.uk/blog/?p=3403

**Comment by trivialfis** [comment]
@hcho3 Thanks for the link!  That's useful.

**Comment by trivialfis** [comment]
@hcho3 This should be ready now if the tests don't go nuts.  Currently I can't write the documentation for ranking #4039 .

**Comment by hcho3** [comment]
@trivialfis Okay, I'll review.

File: src/common/hist_util.cc [hunk 1]
```diff
@@ -35,30 +35,47 @@ void HistCutMatrix::Init(DMatrix* p_fmat, uint32_t max_num_bins) {
...
       CHECK_EQ(nthread, omp_get_num_threads());
       auto tid = static_cast<unsigned>(omp_get_thread_num());
       unsigned begin = std::min(nstep * tid, ncol);
       unsigned end = std::min(nstep * (tid + 1), ncol);
+
+      // Data groups, used in ranking.
+      std::vector<bst_uint> const& groups = info.group_ptr_;
+      size_t const num_groups = groups.size() == 0 ? 0 : groups.size() - 1;
+      size_t group_ind =  // index into groups
+          tid * static_cast<unsigned>((num_groups + nthread - 1) / nthread);
```
Comment (hcho3) [inline src/common/hist_util.cc]: I don't think this will work when `p_fmat->GetRowBatches()` returns multiple batches.

**Comment by hcho3** [comment]
@trivialfis Can you look at my latest revision and see if it's good?

**Comment by hcho3** [comment]
Oops, I think it's wrong. Let me try again

**Comment by trivialfis** [comment]
@hcho3 Do we need to calculate different `group_ind` for each thread since they might be processing data belonging to different groups?  And, it might be worthy to use some sort of binary search to find the index instead of a linear search.

**Comment by trivialfis** [comment]
I can try it again tomorrow unless you beat me to it. :)

**Comment by hcho3** [comment]
@trivialfis Notice that parallelization happens over columns, not rows

**Comment by trivialfis** [comment]
@hcho3 I see.  Sorry for the mistake.

**Comment by hcho3** [comment]
@trivialfis No problem. I should probably add another test case with multiple batches.

**Comment by trivialfis** [comment]
@hcho3 That would be great.  I wanted to add some tests for histogram, but it's quite weird since quantile is not well tested.  I don't know how to derive the solution for tests.

File: src/common/hist_util.cc [hunk 2]
```diff
@@ -35,30 +35,62 @@ void HistCutMatrix::Init(DMatrix* p_fmat, uint32_t max_num_bins) {
...
+  std::vector<bst_uint> const& group_ptr = info.group_ptr_;
+  size_t const num_groups = group_ptr.size() == 0 ? 0 : group_ptr.size() - 1;
+  // Use group index for weights?
+  bool const use_group_ind = num_groups != 0 && weights.size() != info.num_row_;
+
   for (const auto &batch : p_fmat->GetRowBatches()) {
-    #pragma omp parallel num_threads(nthread)
+    size_t group_ind;
+    bool group_ind_found = false;
+    if (use_group_ind) {
```
Comment (trivialfis) [inline src/common/hist_util.cc]: Move this test out of for loop.  If we don't use group, then there's no need for this loop.

File: src/common/hist_util.cc [hunk 3]
```diff
@@ -35,30 +35,62 @@ void HistCutMatrix::Init(DMatrix* p_fmat, uint32_t max_num_bins) {
...
+      // Find the group that contains first row of the data batch
+      size_t const base_rowid = batch.base_rowid;
+      for (size_t i = 0; i < num_groups; ++i) {
```
Comment (trivialfis) [inline src/common/hist_util.cc]: Maybe a binary search is more appropriate since `group_ptr` is sorted by default.

**Comment by hcho3** [comment]
@trivialfis It turns out that currently you cannot enable external memory for fast quantile updater. See #4093. I've added a test anyway, so that we can later add external memory support to fast quantile.

File: src/common/hist_util.cc (hunk 2 above)
Comment (hcho3) [reply src/common/hist_util.cc]: Not sure if we can do that, since this loop will have to run once for every batch (note the use of `batch.base_rowid`.

File: src/common/hist_util.cc (hunk 3 above)
Comment (hcho3) [reply src/common/hist_util.cc]: Can you look at it?

Comment (trivialfis) [reply src/common/hist_util.cc]: Yes.

File: src/common/hist_util.cc (hunk 2 above)
Comment (trivialfis) [reply src/common/hist_util.cc]: I will try it tomorrow, thanks for all the help!

**Comment by hcho3** [comment]
@trivialfis Looks like I uncovered a bug in external memory. I'll remove the failing test and file an new issue.

**Comment by trivialfis** [comment]
@hcho3 It bugs me to add such an overhead.

In the context of using external memory, does the data come in as original order?  If so we can avoid adding such search completely.

**Comment by trivialfis** [comment]
@hcho3 I will keep looking at the external memory part later.  But if you know the answer then please let me know.  The data loading logic is quite confusing to me.  Many thanks!

**Comment by trivialfis** [comment]
@hcho3 I added the monitor for profiling anyway.  I don't think the data will come in as original order.  And since the search performs in O(log n) with log(2^32) == 32, I'm gonna ignore it for now.  Sorry for the noise and ready for another review. :)

**Comment by hcho3** [comment]
@trivialfis For now, it is safe to assume that the external pages are out-of-order. We'll have to dive deeper into external memory when we attempt to fix #4130.

**Comment by trivialfis** [comment]
@hcho3 Could you help taking another look?  I don't think this PR has anything to do with failed R test.

**Comment by trivialfis** [comment]
@hcho3 Mac tests in Travis is not looking good.  Is there anything we can do about it?

**Comment by hcho3** [comment]
@trivialfis This time it’s a bug in Travis CI. Will file an issue today.

**Comment by hcho3** [comment]
@trivialfis Eventually, I hope we find a better alternative for Mac builds, but unfortunately, we cannot get Mac workers with Jenkins since Amazon EC2 doesn't support it.

**Comment by hcho3** [comment]
Filed a support ticket just now. Hope they get back to me soon.

**Comment by hcho3** [comment]
@trivialfis Wow, they got back to me and fixed the issue right away. Builds are working now.

**Comment by trivialfis** [comment]
@hcho3 That's really nice!
